### ISSUE TITLE
Adding roles on unpersisted records (mongoid)

### ISSUE THREAD

**Issue by jthornec** [RolifyCommunity/rolify#423]
Hello,

I've run into an issue with roles on unpersisted records on Mongoid.

**TL;DR**: Adding roles to unpersisted users doesn't work even though it acts like it should. This also causes weird side-effects such as changes to one user's roles affecting another user's roles.

We have a signup process that looks something like this:

```
organization = Organization.new(...)
user = User.new(...)

# ... set up user and organization ...
user.add_role :owner, organization
# ... more setup ...

user.save!
organization.save!
```

Since `add_role` relies on the [shovel (<<) operator](https://github.com/RolifyCommunity/rolify/blob/7bcc986a8c794baf71a358ba5d2699dc1dd4bd40/lib/rolify/adapters/mongoid/role_adapter.rb#L51) to save the association, but the shovel operator [doesn't do the update](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_and_belongs_to_many) if the parent object (`user`) isn't persisted, the `role.users` side of the relationship doesn't get saved. Later, `user.save!` saves the change on the `user.roles` side... but it doesn't save the `role` record, and so things go out of sync.

Essentially, this happens:
```
user.roles.first.users
=> []
```

Note that `user.has_role?` works, because it checks via the `user.roles` side of the relationship, which is set up and persisted.

This gets really tricky when the `remove_role` [cleanup code](https://github.com/RolifyCommunity/rolify/blob/34ed256cd8bd904810fa05bb75e6f438a1a3cb8c/lib/rolify/adapters/mongoid/role_adapter.rb#L76-L78) runs, because it causes the following behaviour:

```
user.has_role? :owner, organization
=> true

another_user = User.find [...]
another_user.add_role :owner, organization
another_user.remove_role :owner, organization

user.has_role? :owner, organization
=> false
```

Specifically, changing the roles on a different user caused our first user to lose ownership of the organization, because the organization's `:owner` `role` record got destroyed, because that record's `.users` was `[]`.

----

I can think of a few different ways to fix this:

1. If adding roles on unpersisted records isn't supported, `add_role` could check `persisted?` on the user and/or resource and raise an exception (or log a warning) to prevent us from thinking that this should work :)
2. `add_role` could explicitly `save` the role, possibly behind a check for `changed?`. Since the `Role` is already guaranteed to be persisted during this operation (via `find_or_create`), `save`ing it shouldn't cause any unexpected surprises.
3. The `rolify` call in the `User` model could set up the `roles` relationship to include something like `autosave: true`, so the role gets saved when we do `user.save!`.

I'd be happy to put together a PR with any of those solutions (and the accompanying specs), but I'm not sure which path would make the most sense, and there might be other options that I haven't considered.

Thanks!
